pkg/sources/docker/docker.go

```diff
@@ -0,0 +1,220 @@
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/source_metadatapb"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/sourcespb"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/sources"
+)
+
+type Source struct {
+	name     string
+	sourceId int64
+	jobId    int64
+	verify   bool
+	conn     sourcespb.Docker
+	sources.Progress
+}
+
+var FilesizeLimitBytes int64 = 10 * 1024 * 1024 // 10MB
+
+// Ensure the Source satisfies the interface at compile time.
+var _ sources.Source = (*Source)(nil)
+
+// Type returns the type of source.
+// It is used for matching source types in configuration and job input.
+func (s *Source) Type() sourcespb.SourceType {
+	return sourcespb.SourceType_SOURCE_TYPE_DOCKER
+}
+
+func (s *Source) SourceID() int64 {
+	return s.sourceId
+}
+
+func (s *Source) JobID() int64 {
+	return s.jobId
+}
+
+// Init initializes the source.
+func (s *Source) Init(_ context.Context, name string, jobId, sourceId int64, verify bool, connection *anypb.Any, concurrency int) error {
+	s.name = name
+	s.sourceId = sourceId
+	s.jobId = jobId
+	s.verify = verify
+
+	// Reset metrics for this source at initialization time.
+	dockerImagesScanned.WithLabelValues(s.name).Set(0)
+	dockerLayersScanned.WithLabelValues(s.name).Set(0)
+
+	if err := anypb.UnmarshalTo(connection, &s.conn, proto.UnmarshalOptions{}); err != nil {
+		return fmt.Errorf("error unmarshalling connection: %w", err)
+	}
+
+	return nil
+}
+
+// Chunks emits data over a channel that is decoded and scanned for secrets.
+func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk) error {
+	remoteOpts, err := s.remoteOpts()
+	if err != nil {
+		return err
+	}
+
+	for _, image := range s.conn.GetImages() {
+		var img v1.Image
+		var err error
+		var base, tag string
+
+		if strings.HasPrefix(image, "file://") {
+			image = strings.TrimPrefix(image, "file://")
+			base = image
+			img, err = tarball.ImageFromPath(image, nil)
+			if err != nil {
+				return err
+			}
+		} else {
+			base, tag = baseAndTagFromImage(image)
+			imageName, err := name.NewTag(image)
+			if err != nil {
+				return err
+			}
+
+			img, err = remote.Image(imageName, remoteOpts...)
+			if err != nil {
+				return err
+			}
+		}
+
+		layers, err := img.Layers()
+		if err != nil {
+			return err
+		}
+
+		for _, layer := range layers {
+			digest, err := layer.Digest()
+			if err != nil {
+				return err
+			}
+
+			rc, err := layer.Compressed()
+			if err != nil {
+				return err
+			}
+
+			defer rc.Close()
+
+			gzipReader, err := gzip.NewReader(rc)
+			if err != nil {
+				return err
+			}
+
+			defer gzipReader.Close()
+
+			tarReader := tar.NewReader(gzipReader)
+
+			for {
+				header, err := tarReader.Next()
+				if err == io.EOF {
+					break // End of archive
+				}
+				if err != nil {
+					return err
+				}
+
+				// Skip files larger than FilesizeLimitBytes
+				if header.Size > FilesizeLimitBytes {
+					continue
+				}
+
+				file := bytes.NewBuffer(nil)
```

Comment:
**Nit:** We know the size, so we can initialize the buffer with a capacity:

```go
file := bytes.NewBuffer(make([]byte, 0, header.Size))
```

Reply:
Great suggestion